src/main/java/gregtech/api/recipes/ingredients/GTRecipeItemInput.java

```diff
@@ -0,0 +1,163 @@
+package gregtech.api.recipes.ingredients;
+
+import gregtech.api.recipes.ingredients.NBTMatching.NBTMatcher;
+import gregtech.api.recipes.ingredients.NBTMatching.NBTcondition;
+import net.minecraft.item.ItemStack;
+import net.minecraft.nbt.NBTTagCompound;
+import net.minecraftforge.fluids.FluidStack;
+
+import java.lang.ref.WeakReference;
+import java.util.Objects;
+
+public class GTRecipeItemInput implements IGTRecipeInput {
+    int amount;
+    boolean isConsumable = true;
+    NBTMatcher nbtMatcher;
+    NBTcondition nbtCondition;
+    ItemStack inputStacks;
+
+    protected GTRecipeItemInput(ItemStack stack) {
+        this.inputStacks = stack;
+        amount = stack.getCount();
+    }
+
+    protected GTRecipeItemInput(ItemStack stack, int amount) {
+        this.inputStacks = stack;
+        this.amount = amount;
+    }
+
+    public static GTRecipeItemInput getOrCreate(ItemStack stack, int amount) {
+        stack.setCount(amount);
+        return getFromCache(new GTRecipeItemInput(stack));
+    }
+
+    private static GTRecipeItemInput getFromCache(GTRecipeItemInput realIngredient) {
+        if (GTIngredientCache.INSTANCES.get(realIngredient) == null) {
+            GTIngredientCache.INSTANCES.put(realIngredient, new WeakReference<>(realIngredient));
+        } else {
+            realIngredient = GTIngredientCache.INSTANCES.get(realIngredient).get();
+        }
+        return realIngredient;
+    }
+
+    public static IGTRecipeInput getOrCreate(IGTRecipeInput ri, int i) {
+        return getFromCache(new GTRecipeItemInput(ri.getInputStack(), i));
+    }
+
+    public static IGTRecipeInput getOrCreate(IGTRecipeInput ri) {
+        return getFromCache(new GTRecipeItemInput(ri.getInputStack()));
+    }
+
+    public static IGTRecipeInput getOrCreate(ItemStack ri) {
+        return getFromCache(new GTRecipeItemInput(ri));
+    }
+
+    @Override
+    public int getAmount() {
+        return this.amount;
+    }
+
+    @Override
+    public GTRecipeItemInput setNonConsumable() {
+        GTRecipeItemInput ret = new GTRecipeItemInput(this.inputStacks, this.amount);
+        ret.isConsumable = false;
+        if (GTIngredientCache.NON_CONSUMABLE_INSTANCES.get(ret) == null) {
+            GTIngredientCache.NON_CONSUMABLE_INSTANCES.put(ret, new WeakReference<>(ret));
+        } else {
+            ret = GTIngredientCache.NON_CONSUMABLE_INSTANCES.get(ret).get();
+        }
```

Comment:
Why is this the only Ingredient that gets added to the Non-consumable map. Neither the Ore inputs nor the fluid inputs get added.